libraries/appbase/include/appbase/plugin.hpp

```diff
@@ -17,58 +17,151 @@
 
 namespace appbase {
 
-   using boost::program_options::options_description;
-   using boost::program_options::variables_map;
-   using std::string;
-   using std::vector;
-   using std::map;
-
-   class application;
-   application& app();
-
-   class abstract_plugin {
-      public:
-         enum state {
-            registered, ///< the plugin is constructed but doesn't do anything
-            initialized, ///< the plugin has initlaized any state required but is idle
-            started, ///< the plugin is actively running
-            stopped ///< the plugin is no longer running
-         };
-
-         virtual ~abstract_plugin(){}
-
-         virtual state get_state()const = 0;
-         virtual const std::string& get_name()const  = 0;
-         virtual void set_program_options( options_description& cli, options_description& cfg ) = 0;
-         virtual void register_dependencies() = 0;
-         virtual void initialize(const variables_map& options) = 0;
-         virtual void startup() = 0;
-         virtual void shutdown() = 0;
-
-      protected:
-         typedef std::function<void(abstract_plugin&)> plugin_processor;
-
-         /** Abstract method to be reimplemented in final plugin implementation.
-             It is a part of initialization/startup process triggerred by main application.
-             Allows to process all plugins, this one depends on.
-         */
-         virtual void plugin_for_each_dependency(plugin_processor&& processor) = 0;
-
-         /** Abstract method to be reimplemented in final plugin implementation.
-             It is a part of initialization process triggerred by main application.
-         */
-         virtual void plugin_initialize( const variables_map& options ) = 0;
-         /** Abstract method to be reimplemented in final plugin implementation.
-             It is a part of startup process triggerred by main application.
-         */
-         virtual void plugin_startup() = 0;
-         /** Abstract method to be reimplemented in final plugin implementation.
-             It is a part of shutdown process triggerred by main application.
-         */
-         virtual void plugin_shutdown() = 0;
+using boost::program_options::options_description;
+using boost::program_options::variables_map;
+using std::string;
+using std::vector;
+using std::map;
 
+class application;
+
+class abstract_plugin : public std::enable_shared_from_this<abstract_plugin>
+{
+public:
+   enum state {
```

Comment:
enum class ?

Reply:
Why used scoped enums? C++11+ shall handle enscoped enums boundaries checkings the same way, just syntactically it is slightly more comples to use scoped ones compared to unscoped.

Reply:
because unscoped enums are source of bugs. They exist in global scope(generally speaking), which is basically something that you want to avoid because it results in name conflicts and there is implicit conversion to integer, which might not be what you want... And if you do, you can use static_cast...

Reply:
But this one is scoped by the namespace and abstract plugin class :)

Reply:
i know, still a nice and recommended habit not to use unscoped enums in post-11 C++ at all :D